validator.go

```diff
@@ -0,0 +1,202 @@
+package jwt
+
+import (
+	"crypto/subtle"
+	"fmt"
+	"time"
+)
+
+// Validator is the core of the new Validation API. It is
+type Validator struct {
+	// leeway is an optional leeway that can be provided to account for clock skew.
+	leeway time.Duration
+
+	// timeFunc is used to supply the current time that is needed for
+	// validation. If unspecified, this defaults to time.Now.
+	timeFunc func() time.Time
+
+	// verifyIat specifies whether the iat (Issued At) claim will be verified.
+	// According to https://www.rfc-editor.org/rfc/rfc7519#section-4.1.6 this
+	// only specifies the age of the token, but no validation check is
+	// necessary. However, if wanted, it can be checked if the iat is
+	// unrealistic, i.e., in the future.
+	verifyIat bool
+
+	// expectedAud contains the audiences this token expects. Supplying an empty
+	// string will disable aud checking.
+	expectedAud string
```

Comment:
As mentioned, checking the issuer is up to the application (such as an OpenID connect client), since it is only meaningful to the application. I can check whether it makes sense to include a simple string comparison, similar to the audience.

Reply:
> the JWKS repo 

I assume this is referencing [`keyfunc`](https://github.com/MicahParks/keyfunc), but I also have [`jwkset`](https://github.com/MicahParks/jwkset) now :slightly_smiling_face: `jwkset` is currently in pre-release, so there's likely to be breaking changes.

Validating `iss` is not planned for `keyfunc` currently. Without refreshing myself on the project, I think it would be an outlier in the `keyfunc.Options`. If it were included, it likely increase the scope of the project to other things like `aud` and `sub` at least. If anyone feels it should be in the project, feel free to open an issue. I'll take more time to consider it.

Currently, the `jwkset` package does not have any sort of client, so any validation of `aud` would be in the distant future.

> But in the end, the application must check if the issuer corresponds to the key and as this is application specific, we probably cannot validate this in the library.

Perhaps there's something I'm missing here that would invalidate my upcoming suggestion. RFC 7519 notes for both `iss` and `aud` that they are "generally application specific". With that, I think they both have an equal argument to be included in `Valdiator` options. It seems to me that since the `Valdiator` options can be given by the package user, the "application specific" behavior can be applied through this new API, if the options are present.

* `iss`: https://www.rfc-editor.org/rfc/rfc7519#section-4.1.1
* `aud`: https://www.rfc-editor.org/rfc/rfc7519#section-4.1.3

Now that it's been mentioned, I think that _both_ `iss` and `sub` should be included in this PR as options in the same way that `aud` currently is. I know it's technically feasible for a package user to check those manually post JWT validation, but if `aud` is included in the `Validator`, I'm under the opinion that **all standardized claims** that might be checked for JWT validity should have an option, including ones that aren't mentioned in this thread. This isn't to mention the required validity checks, like `exp`.

I think it might also be a good idea to make an exported Go data structure that contains all `Validator` options like `iat`, `aud`, leeway, timeFunc, etc. This data structure would have an associated function to apply the options to the `Validator`. This would help users identify options for `Validator` without finding needing to look for all functions that return have a return type of `ValidatorOption`.

This data structure could gain more fields over time.

Something like this, perhaps:
```go
type ValidatorOptions struct {
	ExpectedAudience string
	Leeway           time.Duration
	TimeFunc         func() time.Time
	ValidateIssuedAt bool
}

func WithOptions(options ValidatorOptions) ValidatorOption {
	return func(v *Validator) {
		v.expectedAud = options.ExpectedAudience
		v.leeway = options.Leeway
		v.timeFunc = options.TimeFunc
		v.verifyIat = options.ValidateIssuedAt
	}
}
```

Reply:
> > the JWKS repo
> 
> I assume this is referencing [`keyfunc`](https://github.com/MicahParks/keyfunc), but I also have [`jwkset`](https://github.com/MicahParks/jwkset) now 🙂 `jwkset` is currently in pre-release, so there's likely to be breaking changes.

Yeah I meant `keyfunc`. :)

> 
> Validating `iss` is not planned for `keyfunc` currently. Without refreshing myself on the project, I think it would be an outlier in the `keyfunc.Options`. If it were included, it likely increase the scope of the project to other things like `aud` and `sub` at least. If anyone feels it should be in the project, feel free to open an issue. I'll take more time to consider it.
> 
> Currently, the `jwkset` package does not have any sort of client, so any validation of `aud` would be in the distant future.
> 
> > But in the end, the application must check if the issuer corresponds to the key and as this is application specific, we probably cannot validate this in the library.
> 
> Perhaps there's something I'm missing here that would invalidate my upcoming suggestion. RFC 7519 notes for both `iss` and `aud` that they are "generally application specific". With that, I think they both have an equal argument to be included in `Valdiator` options. It seems to me that since the `Valdiator` options can be given by the package user, the "application specific" behavior can be applied through this new API, if the options are present.
> 
> * `iss`: https://www.rfc-editor.org/rfc/rfc7519#section-4.1.1
> * `aud`: https://www.rfc-editor.org/rfc/rfc7519#section-4.1.3
> 
> Now that it's been mentioned, I think that _both_ `iss` and `sub` should be included in this PR as options in the same way that `aud` currently is. I know it's technically feasible for a package user to check those manually post JWT validation, but if `aud` is included in the `Validator`, I'm under the opinion that **all standardized claims** that might be checked for JWT validity should have an option, including ones that aren't mentioned in this thread. This isn't to mention the required validity checks, like `exp`.

I can definitely include an "require IAT/ISS" similar to aud, but the issue is that the [rfc8725](https://datatracker.ietf.org/doc/html/rfc8725#section-3.8) actually suggest this keybinding validation, for example in OpenID connect by looking up the JWKS in the issuer URL; and this is something which I do not see in this library.

> 
> I think it might also be a good idea to make an exported Go data structure that contains all `Validator` options like `iat`, `aud`, leeway, timeFunc, etc. This data structure would have an associated function to apply the options to the `Validator`. This would help users identify options for `Validator` without finding needing to look for all functions that return have a return type of `ValidatorOption`.
> 
> This data structure could gain more fields over time.
> 
> Something like this, perhaps:
> 
> ```go
> type ValidatorOptions struct {
> 	ExpectedAudience string
> 	Leeway           time.Duration
> 	TimeFunc         func() time.Time
> 	ValidateIssuedAt bool
> }
> 
> func WithOptions(options ValidatorOptions) ValidatorOption {
> 	return func(v *Validator) {
> 		v.expectedAud = options.ExpectedAudience
> 		v.leeway = options.Leeway
> 		v.timeFunc = options.TimeFunc
> 		v.verifyIat = options.ValidateIssuedAt
> 	}
> }
> ```

I am a bit torn with this idea, because I think there is only a limited set of use-cases which require you to set *all* values. Instead we should rather work on a proper README / documentation of the API; something which I any way wanted to tackle: Redesign the current README and include short examples in there.

Reply:
There is one more fuzzy detail: The `aud`, `iss`, `sub` claims are all OPTIONAL according to the JWT RFC. However, if I as a developer use `WithAudience` / `WithIssuer` / `WithSubject` I guess I would *assume* that the claim is there (aka REQUIRED), and otherwise fail, correct? I think this behaviour should be ok from the RFC side. The question would be: does that make sense? Should we make it configurable?

Reply:
> for example in OpenID connect by looking up the JWKS in the issuer URL; and this is something which I do not see in this library.

I agree that this example seems out of scope for this project.

> Instead we should rather work on a proper README / documentation of the API

This accomplishes the same goal as my suggestion, so it sounds good to me :slightly_smiling_face:

Reply:
> There is one more fuzzy detail: The `aud`, `iss`, `sub` claims are all OPTIONAL according to the JWT RFC. However, if I as a developer use `WithAudience` / `WithIssuer` / `WithSubject` I guess I would _assume_ that the claim is there (aka REQUIRED), and otherwise fail, correct? I think this behaviour should be ok from the RFC side. The question would be: does that make sense? Should we make it configurable?

If I were a package user and I included `WithAudience` I would expect there to be an error if `aud` was not present. But maybe I'm poisoned with knowledge.

Reply:
> Should we make it configurable?

I will suggest to make the REQUIRED presence when `WithAudience` / `WithIssuer` / `WithSubject` is set and not to let it be configurable. I think that would be best for _most use_ cases.

If a package user has a use case where it could be optional, I believe each of these claims can be checked after `Validator` is finished.

> I think that would be best for _most use_ cases.

I may be making too much of an assumption here, so anyone can contradict that.

Reply:
I agree 100 % @MicahParks. Then I'll implement the remaining `WithXXX`... Maybe another idea that comes into mind. While the string comparison functions are pretty much straightforward, at least for `subject` maybe it makes sense to have something like `WithSubjectPattern`; because you might want to validate the subject according to some business logic. On the other hand this could easily be done in the `CustomValidation` of your own claim. Not sure if this would be over-engineered.

Reply:
> > Maybe another idea that comes into mind. While the string comparison functions are pretty much straightforward, at least for `subject` maybe it makes sense to have something like `WithSubjectPattern`; because you might want to validate the subject according to some business logic. On the other hand this could easily be done in the `CustomValidation` of your own claim. Not sure if this would be over-engineered.
> 
> I, personally, would leave that out of the initial PR. If there's a good use case for it, it could be added in the future if prompted by an issue.

I'll keep it in a separate branch/PR then. I did a quick hack of it and I quite like it. Looks like this:
```
claims := RegisteredClaims{Subject: "MyUser"}
NewValidator(
  WithSubjectPattern(HasPrefix("My")),
).Validate(claims)
```